File: common/src/main/java/org/bread_experts_group/breadlib/task/render/LayeredDrawTask.java
package org.bread_experts_group.breadlib.task.render;

import net.minecraft.client.gui.LayeredDraw;
import net.minecraft.resources.ResourceLocation;
import org.bread_experts_group.breadlib.task.Task;

import java.util.HashMap;
import java.util.Map;

// todo order based layer adding for all three platforms (render our layers behind or in front of other layers)
public class LayeredDrawTask extends Task {
	private final Map<ResourceLocation, LayeredDraw.Layer> layers = new HashMap<>();

	public Map<ResourceLocation, LayeredDraw.Layer> getLayers() {
		return layers;
	}

	public void add(ResourceLocation location, LayeredDraw.Layer layer) {
		this.layers.put(location, layer);
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/task/render/LayeredDrawTask.kt
package org.bread_experts_group.breadlib.task.render

import net.minecraft.client.gui.LayeredDraw
import net.minecraft.resources.ResourceLocation
import org.bread_experts_group.breadlib.task.Task

// todo layer based logic currently does not exist on fabric, refer to MixinGui
class LayeredDrawTask : Task() {
	@JvmField
	val layers: MutableList<LayerInfo> = mutableListOf()

	enum class Ordering { BEFORE, AFTER }
	data class LayerInfo(
		val id: ResourceLocation,
		val other: ResourceLocation?,
		val order: Ordering,
		val layer: LayeredDraw.Layer
	)

	fun getLayers(): Collection<LayerInfo> = this.layers

	fun addAbove(id: ResourceLocation, other: ResourceLocation, layer: LayeredDraw.Layer) {
		this.layers.add(LayerInfo(id, other, Ordering.AFTER, layer))
	}

	fun addBelow(id: ResourceLocation, other: ResourceLocation, layer: LayeredDraw.Layer) {
		this.layers.add(LayerInfo(id, other, Ordering.BEFORE, layer))
	}

	fun addAboveAll(id: ResourceLocation, layer: LayeredDraw.Layer) {
		this.layers.add(LayerInfo(id, null, Ordering.AFTER, layer))
	}

	fun addBelowAll(id: ResourceLocation, layer: LayeredDraw.Layer) {
		this.layers.add(LayerInfo(id, null, Ordering.BEFORE, layer))
	}
}